File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Tests/Api/BookstoreServiceTest.cs
using System.Collections.Generic;
using System.Net;
using FluentAssertions;
using NUnit.Extension.DependencyInjection;
using RekindleTaf.Controllers;
using RekindleTaf.Exceptions;
using RekindleTaf.Model;
using RekindleTaf.Model.Factory;
using RekindleTaf.Utils;

namespace RekindleTaf.Tests.Api;

[DependencyInjectingTestFixture]
public class BookstoreServiceTest(IFactoryKit factoryKit, IBookstoreController bookstoreController) : Base
{
    private readonly List<Bookstore> Bookstores = new();
    
    [OneTimeTearDown]
    public void Cleanup()
    {
        if (Bookstores.Count == 0) return;
        Bookstores.ForEach(bookstoreController.DeleteSuccessfully);
    }

    [Test]
    public void Create_NewBookstore_Should_Return_OK()
    {
        var bookstore = factoryKit.Factory<Bookstore>(FactoryKey.Bookstore).Create(BookstoreKey.StdBookstoreActive);
        bookstore = bookstoreController.PostSuccessfully(bookstore);
        Bookstores.Add(bookstore);
        bookstore.Id.Should().NotBeEmpty();
    }
    
    [Test]
    public void Get_NewBookstore_Should_Return_OK()
    {
        var bookstore = factoryKit.Factory<Bookstore>(FactoryKey.Bookstore).Create(BookstoreKey.StdBookstoreActive);
        bookstore = bookstoreController.PostSuccessfully(bookstore);
        Bookstores.Add(bookstore);
        var bookstoreFetched = bookstoreController.GetSuccessfully(bookstore.Id!);
        bookstoreFetched.Should().BeEquivalentTo(bookstore);
    }
    
    [Test]
    public void Get_AllBookstores_Should_Return_OK()
    {
        var bookstoresFetched = bookstoreController.GetAllSuccessfully();
        bookstoresFetched.Should().NotBeEmpty();
    }
    
    [Test]
    public void Update_Bookstore_Should_Return_NoContent()
    {
        var bookstore = factoryKit.Factory<Bookstore>(FactoryKey.Bookstore).Create(BookstoreKey.StdBookstoreActive);
        bookstore = bookstoreController.PostSuccessfully(bookstore);
        Bookstores.Add(bookstore);
        var updatedName = "UpdatedName";
        bookstore.Name = updatedName;
        bookstoreController.PutSuccessfully(bookstore);
        var bookstoreFetched = bookstoreController.GetSuccessfully(bookstore.Id!);
        bookstoreFetched.Should().BeEquivalentTo(bookstore);
    }
    
    [Test]
    public void Delete_Bookstore_Should_Return_NoContent()
    {
        var bookstore = factoryKit.Factory<Bookstore>(FactoryKey.Bookstore).Create(BookstoreKey.StdBookstoreActive);
        bookstore = bookstoreController.PostSuccessfully(bookstore);
        bookstoreController.DeleteSuccessfully(bookstore);
        var response = bookstoreController.Get(bookstore.Id!).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.BookstoreNotFound);
    }
}

File: typescript-project/rekindle-book-store-aqa/tests/bookstore-service.test.ts
import {beforeAll, afterAll, describe, expect, test} from '@jest/globals';
import {Bookstore} from '@/model/domain-model';
import {Factory, FactoryKey, FactoryKit} from '@/model/factory';
import {BookstoreKey} from '@/model/bookstore-factory';
import {HttpStatusCode} from 'axios';
import {ErrorMessage} from '../src/utils/utils';
import {container} from 'tsyringe';
import {RestClient} from '@/rest/rekindle-rest-client';
import {BookstoreCrudController} from "@/rest/controllers/rest-controllers";

describe('Bookstore Service Test', () => {
    let restClient: RestClient;
    let factoryKit: FactoryKit;
    let bookstoreController: BookstoreCrudController;
    let bookstoreFactory: Factory<Bookstore>;
    let bookstores: Bookstore[] = [];

    beforeAll(async () => {
        restClient = container.resolve('RestClient');
        factoryKit = container.resolve('FactoryKit');
        bookstoreController = container.resolve('BookstoreCrudController');
        bookstoreFactory = factoryKit.factory<Bookstore>(FactoryKey.bookstore);
        await restClient.setupToken();
    });

    afterAll(async () => {
        if (bookstores.length === 0) return;
        for (const bookstore of bookstores) {
            bookstoreController.deleteSuccessfully(bookstore);
        }
    });

    test('Create New Bookstore Should Return OK', async () => {
        let bookstore = bookstoreFactory.create(BookstoreKey.stdBookstoreActive);
        bookstore = await bookstoreController.postSuccessfully(bookstore);
        bookstores.push(bookstore);
        expect(bookstore.id).toBeDefined();
    });

    test('Get New Bookstore Should Return OK', async () => {
        let bookstore = bookstoreFactory.create(BookstoreKey.stdBookstoreActive);
        bookstore = await bookstoreController.postSuccessfully(bookstore);
        bookstores.push(bookstore);
        const bookstoreFetched = await bookstoreController.getSuccessfully(bookstore.id!);
        expect(bookstoreFetched).toEqual(bookstore);
    });

    test('Get All Bookstores Should Return OK', async () => {
        const bookstoresFetched = await bookstoreController.getAllSuccessfully();
        expect(bookstoresFetched).not.toHaveLength(0);
    });

    test('Update Bookstore Should Return No Content', async () => {
        let bookstore = bookstoreFactory.create(BookstoreKey.stdBookstoreActive);
        bookstore = await bookstoreController.postSuccessfully(bookstore);
        bookstores.push(bookstore);
        const updatedName = 'UpdatedName';
        bookstore.name = updatedName;
        await bookstoreController.putSuccessfully(bookstore);
        const bookstoreFetched = await bookstoreController.getSuccessfully(bookstore.id);
        expect(bookstoreFetched.name).toBe(updatedName);
    });

    test('Delete Bookstore Should Return No Content', async () => {
        let bookstore = bookstoreFactory.create(BookstoreKey.stdBookstoreActive);
        bookstore = await bookstoreController.postSuccessfully(bookstore);
        await bookstoreController.deleteSuccessfully(bookstore);
        await bookstoreController.get(bookstore.id!)
            .catch((error) => {
                expect(error.response.status).toBe(HttpStatusCode.NotFound);
                expect(error.response.data.errorMessage).toBe(ErrorMessage.bookstoreNotFound);
            });
    });
});